### DISCUSSION TITLE
params_filter introspection

### DISCUSSION THREAD

**Discussion by Spone** [palkan/action_policy#213] [Q&A] [answered]
I'm using the [ActionController parameters](https://actionpolicy.evilmartians.io/#/scoping?id=action-controller-parameters) feature. I'd like to be able to query which params are allowed to the current user, in order to decide which form fields to display in my admin panel.

With Pundit I would do something like:

```rb
- if @post.permitted_attributes.include?(:title)
  = f.text_field :title
```

What would be the easiest way to achieve this with ActionPolicy? Thanks!

**Comment by palkan** [comment, marked as answer]
I would suggest using a plain old Ruby method in the policy class for that:

```ruby
class PostPolicy < ApplicationPolicy
  def permitted_attributes
    %i[title]
  end

  def permitted_attribute?(name)
    permitted_attributes.include?(name)
  end

  # you can re-use this method in the params scope
  params_filter { _1.permit(*permitted_attributes)  }
end

# in your view
- if policy_for(@post).permitted_attribute?(:title)
  = f.text_field :title 
```

Alternatively, you can use scopes (if you need different contexts):

```ruby
class UserPolicy < ApplicationPolicy
  scope_for :permitted_params do
    %i[name email]
  end

  scope_for :permitted_params, :update do
    %i[name]
  end
end

authorized_scope(user, type: :permitted_params)

authorized_scope(user, type: :permitted_params, as: :update)
```

This variant is more verbose (and doesn't reads well), though could be wrapped into a custom API.

**Reply by Spone** [reply]
Thanks for the advice :)

**Reply by Spone** [reply]
Just a quick addition: in order to be able to call `policy_for` from a view, `helper_method :policy_for` must be added to the controller.

Also, there was a typo in the view code:

```diff
- if policy_for(@post).permitted_attribute?(:title)
+ if policy_for(record: @post).permitted_attribute?(:title)
```
